### PR TITLE
drivers: gpio: move non-standard dts flags to be soc specific

### PR DESCRIPTION
**Pull request by henrikbrixandersen**
Reserve the upper 8 bits of `gpio_dt_flags_t` for SoC specific flags and move the non-standard, hardware-specific GPIO devicetree flags (IO voltage level, drive strength, debounce filter) from the generic `dt-bindings/gpio/gpio.h` header to SoC specific dt-bindings headers.
    
Some of the SoC specific dt-bindings flags take up more bits than necessary in order to retain backwards compatibility with the deprecated GPIO flags. The width of these fields can be reduced/optimized once the deprecated flags are removed.
    
Remove hardcoded use of `GPIO_INT_DEBOUNCE` in GPIO client drivers. This flag can now be set in the devicetree for boards/SoCs with debounce filter support. The SoC specific debounce flags have had the `_INT` part of their name removed since these flag must be passed to `gpio_pin_configure()`, not `gpio_pin_interrupt_configure()`.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

### CODE REVIEW

File: include/dt-bindings/gpio/gpio.h [hunk 1]
```diff
@@ -102,6 +102,59 @@
...
+ * The interface supports two different drive strengths:
+ * `DFLT` - The lowest drive strength supported by the HW
+ * `ALT` - The highest drive strength supported by the HW
+ *
+ * On hardware that supports only one standard drive strength, both
+ * `DFLT` and `ALT` have the same behavior.
+ * @{
+ */
+/** @cond INTERNAL_HIDDEN */
+#define GPIO_DS_LOW_POS 20
```
Comment (keith-zephyr) [inline include/dt-bindings/gpio/gpio.h:128]: gpio_dt_flags_t is currently defined as uint8_t.  gpio_dt_flags_t should be updated as part of this change.  

If the GPIO flags in devicetree are going to be extended past 8 bits, It probably makes sense to just eliminate gpio_dt_flags_t and only use gpio_flags_t throughout Zephyr.

**Review by keith-zephyr — DISMISSED** [review]
If the dt-bindings are going to be extended, I think all the GPIO flags should be moved into the dt-bindings instead of just the drive strength flags.

File: include/dt-bindings/gpio/gpio.h [hunk 2]
```diff
@@ -102,6 +102,59 @@
 
 /** @} */
 
+/**
```
Comment (galak) [inline include/dt-bindings/gpio/gpio.h:105]: Add comment about these being zephyr specific like the voltages are.

**Comment by carlescufi** [comment]
@henrikbrixandersen ping

**Comment by keith-zephyr** [comment]
I'd like to see this change get approved.  What's the current status of this PR?

File: include/dt-bindings/gpio/gpio.h (hunk 2 above)
Comment (henrikbrixandersen) [reply include/dt-bindings/gpio/gpio.h:105]: Added, thanks.

**Comment by henrikbrixandersen** [comment]
> I'd like to see this change get approved. What's the current status of this PR?

@keith-zephyr Thanks for reminding me. I have updated this PR to extend the size of `gpio_dt_flags_t` from 8 to 16 bits and moved the flags accordingly.

File: include/dt-bindings/gpio/gpio.h (hunk 1 above)
Comment (henrikbrixandersen) [reply include/dt-bindings/gpio/gpio.h]: I have opted to extend `gpio_dt_flags_t` to 16 bits, but kept it for now. I agree that we should consider removing it, but I would like to keep that out of this PR.

Comment (keith-zephyr) [reply include/dt-bindings/gpio/gpio.h]: We (Google) have a use case for specifying the GPIO_INPUT/OUTPUT and the GPIO_OUTPUT_INIT_LOW/INIT_HIGH flags in the devicetree. So I'm in favor of extended gpio_dt_flags_t to 32 bits and making all the flags available in the bindings header.

Comment (henrikbrixandersen) [reply include/dt-bindings/gpio/gpio.h]: That would make all of our GPIO flags Zephyr-specific devicetree extensions. I am not sure, that's a good plan.

Comment (henrikbrixandersen) [reply include/dt-bindings/gpio/gpio.h]: I will bring this up on todays dev-review meeting.

Comment (henrikbrixandersen) [reply include/dt-bindings/gpio/gpio.h]: We discussed this a bit on todays API meeting, but didn't reach a conclusion.
@mbolivar-nordic Any input on this?

Comment (mbolivar-nordic) [reply include/dt-bindings/gpio/gpio.h]: > @mbolivar-nordic Any input on this?

I'd like to get @mnkp to approve this. It's always been his design decision as GPIO maintainer that the Zephyr devicetree GPIO flags match Linux's exactly, so I think this is his decision.

**Review by gmarull — DISMISSED** [review]

File: include/dt-bindings/gpio/gpio.h (hunk 1 above)
Comment (carlescufi) [reply include/dt-bindings/gpio/gpio.h]: @henrikbrixandersen could you please comment on the fact that we'd diverge from the Linux GPIO flags in DT?

Comment (henrikbrixandersen) [reply include/dt-bindings/gpio/gpio.h]: @carlescufi That's already present: `The drive strength flags are a Zephyr specific extension ...`

**Comment by mnkp** [comment]
In principle we want to follow [Linux gpio bindings](https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/gpio/gpio.txt) and they don't define a property to set the drive strength in DTS. The drive strength property is defined only in [Linux pinctrl bindings](https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/pinctrl/pincfg-node.yaml) as:
```

  drive-strength:
    $ref: /schemas/types.yaml#/definitions/uint32
    description: sink or source at most X mA
```
It may be better to implement support for setting the drive strength in pinctrl driver and deprecate `GPIO_DS_` flags in gpio driver. Especially that `GPIO_DS_`  flags don't match well hardware capabilities of devices supported by Zephyr.

**Review by anangl — DISMISSED** [review]

**Comment by gmarull** [comment]
> In principle we want to follow [Linux gpio bindings](https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/gpio/gpio.txt?rgh-link-date=2022-02-07T03%3A20%3A32Z) and they don't define a property to set the drive strength in DTS. The drive strength property is defined only in [Linux pinctrl bindings](https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/pinctrl/pincfg-node.yaml?rgh-link-date=2022-02-07T03%3A20%3A32Z) as:
> 
> ```
> 
>   drive-strength:
>     $ref: /schemas/types.yaml#/definitions/uint32
>     description: sink or source at most X mA
> ```
> 
> It may be better to implement support for setting the drive strength in pinctrl driver and deprecate `GPIO_DS_` flags in gpio driver. Especially that `GPIO_DS_` flags don't match well hardware capabilities of devices supported by Zephyr.

The Zephyr GPIO flags do not have units:

```
 * `DFLT` - The lowest drive strength supported by the HW
 * `ALT` - The highest drive strength supported by the HW
```

They really allow setting the lowest/highest value, which seems portable enough and useful for simple cases like e.g., driving LEDs. Drive strength is one of those vendor-specific stuff that can hardly be generalized.

**Comment by henrikbrixandersen** [comment]
> It may be better to implement support for setting the drive strength in pinctrl driver and deprecate `GPIO_DS_` flags in gpio driver. Especially that `GPIO_DS_` flags don't match well hardware capabilities of devices supported by Zephyr.

This doesn't align with the decision of not using pinctrl for GPIO in Zephyr. This was discussed on the API meeting 2021-11-02 (see https://github.com/zephyrproject-rtos/zephyr/pull/39767#issuecomment-957902226).

**Comment by carlescufi** [comment]
API meeting

> * Concern as to there is duplication between pinctrl standard properties and the macros in this Pull Request
> * The pinctrl node is not used for pure GPIOs, hence there is no duplication possible. This is already documented here: https://docs.zephyrproject.org/latest/guides/pinctrl/index.html#pin-control-vs-gpio

- Counterproposal from @mnkp: Reserve the upper 8 bits of `gpio_dt_flags_t` so that they are implementation-dependent. This would be similar to `pinctrl_soc_pin_t` but only for the upper 8-bit of the 16-bit flags. Also remove the current API drive strength and debounce flags.

File: samples/boards/nrf/mesh/onoff_level_lighting_vnd_app/src/app_gpio.c [hunk 3]
```diff
@@ -64,8 +64,7 @@ void app_gpio_init(void)
 
 	button_device[0] = device_get_binding(DT_GPIO_LABEL(SW0_NODE, gpios));
 	gpio_pin_configure(button_device[0], DT_GPIO_PIN(SW0_NODE, gpios),
-			   GPIO_INPUT | GPIO_INT_DEBOUNCE |
```
Comment (mbolivar-nordic) [inline samples/boards/nrf/mesh/onoff_level_lighting_vnd_app/src/app_gpio.c:67]: Removing the debounce flag here seems to change the expected behavior. The sample documents that it works with nrf52840dk_nrf52840, so it seems like this should at least add an overlay for that board specifying the right flags there, no? I have the same question for other similar changes elsewhere in this patch.

Comment (mbolivar-nordic) [reply samples/boards/nrf/mesh/onoff_level_lighting_vnd_app/src/app_gpio.c:67]: Or maybe all the relevant board DTS files need to have the debounce flags added to their button nodes?

File: include/dt-bindings/gpio/andestech-atcgpio100.h [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+#define ZEPHYR_INCLUDE_DT_BINDINGS_GPIO_ANDESTECH_ATCGPIO100_H_
+
+/**
+ * @brief Enable GPIO pin debounce.
+ *
+ * The debounce flag is a Zephyr specific extension of the standard GPIO flags
+ * specified by the Linux GPIO binding. Only applicable for Andestech Technology
+ * Corporation ATCGPIO100 SoCs.
+ */
+#define GPIO_INT_DEBOUNCE (1U << 8)
```
[the remark this answers is not in the document]
Comment (henrikbrixandersen) [reply include/dt-bindings/gpio/andestech-atcgpio100.h]: @gmarull I have updated this PR to namespace the SoC specific GPIO DTS flags.

File: samples/boards/nrf/mesh/onoff_level_lighting_vnd_app/src/app_gpio.c (hunk 3 above)
Comment (henrikbrixandersen) [reply samples/boards/nrf/mesh/onoff_level_lighting_vnd_app/src/app_gpio.c:67]: @mbolivar-nordic That's the thing, really. The sample support `nrf52840dk_nrf52840`, but the nRF GPIO driver does not support `GPIO_INT_DEBOUNCE` at all, so this flag has no effect.

One could argue that all button nodes for boards with a SoC supporting GPIO debouncing should have that flag added, but this is no simple task (60+ boards from what I count) - and some of these may be doing hardware button debouncing, rendering this flag unnecessarily or even counterproductive.

As for the other drivers where I have removed this flag, they are drivers for external LoRa chips where the interrupt lines where configured with debouncing enabled. This should not be needed at all.

**Review by gmarull — DISMISSED** [review]
LGTM. In some cases, like nRF, we can now offer more meaningful options for the users. Will look at it after this gets in.

File: samples/boards/nrf/mesh/onoff_level_lighting_vnd_app/src/app_gpio.c (hunk 3 above)
Comment (mbolivar-nordic) [reply samples/boards/nrf/mesh/onoff_level_lighting_vnd_app/src/app_gpio.c:67]: OK, fair enoguh

**Review by mbolivar-nordic — DISMISSED** [review]
We need an index of device drivers in the documentation that includes information on driver-specific features more than ever with changes like this going in...

File: include/drivers/gpio.h [hunk 5]
```diff
@@ -201,79 +201,6 @@ extern "C" {
 
 /** @} */
 
-/** Enable GPIO pin debounce.
- *
- * @note Drivers that do not support a debounce feature should ignore
- * this flag rather than rejecting the configuration with -ENOTSUP.
- */
-#define GPIO_INT_DEBOUNCE              (1U << 19)
```
Comment (mnkp) [inline include/drivers/gpio.h:209]: We cannot remove the macros, need to deprecate them first, e.g.:
```
#define GPIO_INT_DEBOUNCE              __DEPRECATED_MACRO (1U << 8)
```
The same holds true for drive strength flags.

**Review by mnkp — CHANGES_REQUESTED** [review]
Thanks @henrikbrixandersen, great work! I have some minor but blocking comments. Also, I think we should squash "drivers: gpio: allow specifying GPIO drive strength/debounce flags in dts" commit. I.e. implement all the changes in a single commit.

File: include/drivers/gpio.h (hunk 5 above)
Comment (henrikbrixandersen) [reply include/drivers/gpio.h:209]: It's tricky to both deprecate them and make them SoC specific without overlapping.

**Comment by henrikbrixandersen** [comment]
> Thanks @henrikbrixandersen, great work! I have some minor but blocking comments. Also, I think we should squash "drivers: gpio: allow specifying GPIO drive strength/debounce flags in dts" commit. I.e. implement all the changes in a single commit.

@mnkp Thank you for reviewing. I have squashed to two commits into one. Please see my question on how you'd like to handle deprecation of the old macros.

**Review by brockus-zephyr — DISMISSED** [review]

**Comment by henrikbrixandersen** [comment]
@mnkp Ping?

File: drivers/gpio/gpio_andes_atcgpio100.c [hunk 6]
```diff
@@ -140,7 +141,7 @@ static int gpio_atcgpio100_config(const struct device *port,
 		key = k_spin_lock(&data->lock);
 
 		/* Set de-bounce */
-		if (flags & GPIO_INT_DEBOUNCE) {
+		if (flags & ATCGPIO100_GPIO_INT_DEBOUNCE) {
```
Comment (anangl) [inline drivers/gpio/gpio_andes_atcgpio100.c:144]: I think we should get rid of this misleading `INT_` component from the names of debounce flags. Unlike all other `GPIO_INT_*` flags, the debounce flag is supposed to be used in calls to `gpio_pin_configure()`, not `gpio_pin_interrupt_configure()`.

File: include/drivers/gpio.h (hunk 5 above)
Comment (mnkp) [reply include/drivers/gpio.h:209]: When going the `__DEPRECATED_MACRO` way we would need to make sure that the code compiles and the drivers work just as before. The only difference being the warning message generated by the compiler.

This is possible as long as we use the same bits in the SoC specific flags. I.e. in `include/drivers/gpio.h`
```
#define GPIO_DS_LOW_POS __DEPRECATED_MACRO 9
#define GPIO_DS_HIGH_POS __DEPRECATED_MACRO 11
```
and then in the SoC specific flags e.g. `include/dt-bindings/gpio/nordic-nrf-gpio.h`
```
#define NRF_GPIO_DS_LOW_POS 9
#define NRF_GPIO_DS_HIGH_POS 11
```
We need to keep the deprecated macros for two releases. When they are gone the SoC specific flags can be cleaned up, if required.

File: drivers/gpio/gpio_andes_atcgpio100.c (hunk 6 above)
Comment (henrikbrixandersen) [reply drivers/gpio/gpio_andes_atcgpio100.c]: @anangl Thank you for reviewing. I have removed the `_INT` part of the name for the SoC specific debounce flags.

File: include/drivers/gpio.h (hunk 5 above)
Comment (henrikbrixandersen) [reply include/drivers/gpio.h:209]: @mnkp Thank you for the feedback. I modified this PR to do that.

**Review by gmarull — DISMISSED** [review]

**Review by gmarull — APPROVED** [review]

File: include/dt-bindings/gpio/gpio.h [hunk 7]
```diff
@@ -76,29 +76,26 @@
...
 /** @cond INTERNAL_HIDDEN */
-#define GPIO_VOLTAGE_POS	6
-#define GPIO_VOLTAGE_MASK	(3U << GPIO_VOLTAGE_POS)
+#define GPIO_VOLTAGE_POS        11
+#define GPIO_VOLTAGE_MASK       (3U << GPIO_VOLTAGE_POS)
 /** @endcond */
 
 /** Set pin at the default voltage level */
-#define GPIO_VOLTAGE_DEFAULT	(0U << GPIO_VOLTAGE_POS)
+#define GPIO_VOLTAGE_DEFAULT    (0U << GPIO_VOLTAGE_POS)
```
Comment (anangl) [inline include/dt-bindings/gpio/gpio.h:92]: Shouldn't `__DEPRECATED_MACRO` be added also to `GPIO_VOLTAGE_*` macros?

**Review by anangl — APPROVED** [review]

File: include/dt-bindings/gpio/gpio.h (hunk 7 above)
Comment (henrikbrixandersen) [reply include/dt-bindings/gpio/gpio.h:92]: As far as I can tell, our tools do not support `__DEPRECATED_MACRO` in dts files.

**Review by mnkp — APPROVED** [review]

**Pull request merged**